uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js

```diff
@@ -0,0 +1,591 @@
+
+                describe('change selected page', function () {
+                  var beforeFilteredItems;
+
+                  beforeEach(function () {
+                    beforeFilteredItems = $ctrl.pagination.filteredbreakdown;
+                    $ctrl.pagination.currentPage = 2;
+                    $ctrl.pagination.pageChanged();
+                  });
+
+                  it('should change current page', function () {
+                    expect($ctrl.pagination.currentPage).not.toEqual(1);
+                  });
+
+                  it('should change filtered data', function () {
+                    expect($ctrl.pagination.filteredbreakdown[0]).not.toEqual(beforeFilteredItems[0]);
+                  });
+                });
+              });
+            });
+          });
+        });
+      });
+
+      describe('save leave request', function () {
+
+        describe('with invalid fields', function () {
+
+          it('does not have required fields', function () {
+            expect($ctrl.leaveRequest.from_date).not.toBeDefined();
+          });
+
+          describe('on submit', function () {
+
+            beforeEach(function () {
+              $ctrl.submit();
+              $scope.$digest();
+            });
+
+            it('should fail', function () {
+              expect($ctrl.error).toEqual(jasmine.any(Object));
+            });
+          });
+        });
+
+        describe('with valid fields', function () {
+
+          beforeEach(function () {
+            $ctrl.onDateChange(new Date(), true);
+            $ctrl.onDateChange(new Date(), false);
+            $scope.$digest();
+          });
+
+          it('has all required fields', function () {
+            expect($ctrl.leaveRequest.id).not.toBeDefined();
+            expect($ctrl.leaveRequest.from_date).toBeDefined();
+          });
+
+          describe('on submit', function () {
+
+            beforeEach(function () {
+              spyOn($rootScope, '$emit');
+              $ctrl.submit();
+              $scope.$digest();
+            });
+
+            it('is successful', function () {
+              expect($ctrl.error).not.toEqual(jasmine.any(Object));
+              expect($ctrl.leaveRequest.id).toBeDefined();
+            });
+
+            it('sends event', function () {
+              expect($rootScope.$emit).toHaveBeenCalledWith('LeaveRequest::new', $ctrl.leaveRequest);
+            });
+
+            describe('when balance change is negative', function () {
+
+              beforeEach(function () {
+                $ctrl.selectedAbsenceType = $ctrl.absenceTypes[1];
+                $ctrl.uiOptions.toDate = $ctrl.uiOptions.fromDate = new Date();
+                $ctrl.onDateChange($ctrl.uiOptions.fromDate, true);
+                $ctrl.onDateChange($ctrl.uiOptions.toDate, false);
+                $scope.$digest();
+                $ctrl.submit();
+                $scope.$digest();
+              });
+
+              it('will not save', function () {
```

Comment:
The way tests are structured is just as important as the content of the tests itself, it has a major impact on how easy they are to read
```js
describe('when balance change is negative', function () {


  describe('when the absence type does not allow overuse', function () {
    
    it('does not save', function () {

    });
  });

  describe('when the absence type allows overuse', function () {
    
    it('does save', function () {

    });
  });
});
```